proxy/pkg/zdmproxy/clienthandler.go

```diff
@@ -2241,9 +2244,30 @@ func newFrameProcessor(conf *config.Config, nodeMetrics *metrics.NodeMetrics, co
 	}
 	var mapper StreamIdMapper
 	if connectorType == ClusterConnectorTypeAsync {
-		mapper = NewInternalStreamIdMapper(conf.ProxyMaxStreamIds, streamIdsMetric)
+		mapper = NewInternalStreamIdMapper(maxStreamIds, streamIdsMetric)
 	} else {
-		mapper = NewStreamIdMapper(conf.ProxyMaxStreamIds, streamIdsMetric)
+		mapper = NewStreamIdMapper(maxStreamIds, streamIdsMetric)
 	}
 	return NewStreamIdProcessor(mapper)
 }
+
+// Calculate maximum number of stream IDs. Take the oldest protocol version negotiated between two clusters
+// and apply limit defined in proxy configuration. If origin or target cluster are still running protocol V2,
+// we will limit maximum number of stream IDs to 128 on both clusters. Logic is based on Java driver version 3.x.
+// Java driver 3.x was the last one supporting protocol V2. It establishes control connection first, and then
+// uses negotiated protocol version to configure maximum number of stream IDs on node connections. Driver does NOT
+// change the number of stream IDs on per node basis.
+func maxStreamIds(originProtoVer primitive.ProtocolVersion, targetProtoVer primitive.ProtocolVersion, conf *config.Config) int {
+	maxSupported := maxOutgoingPending
```

Comment:
Where is `maxOutgoingPending` being used? For the channel buffer size? We should keep the channel buffer sizes the same but add a new constant for the max number of stream ids in v3+ and use that constant here. Otherwise if a user tries to set 4000 max stream ids int he conf it won't work even though v3+ supports it.